pkg/kafka/correlation_cache.go

```diff
@@ -0,0 +1,129 @@
+// limitations under the License.
+
+package kafka
+
+import (
+	"time"
+
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+const (
+	requestLifetime = time.Duration(2) * time.Minute
+)
+
+// CorrelationID is the correlation ID in the Kafka protocol
+type CorrelationID int32
+
+type requestsCache map[CorrelationID]*RequestMessage
+
+// CorrelationCache is a cache of requests to correlate responses with requests
+type CorrelationCache struct {
+	cache  requestsCache
+	mutex  lock.RWMutex
+	stopGc chan struct{}
+}
+
+// NewCorrelationCache returns a new correlation cache
+func NewCorrelationCache() *CorrelationCache {
+	cc := &CorrelationCache{
+		cache:  requestsCache{},
+		stopGc: make(chan struct{}),
+	}
+
+	go cc.garbageCollector()
+
+	return cc
+}
+
+// DeleteCache releases the cache and stops the garbage collector. This
+// function must be called when the cache is no longer required, otherwise go
+// routines are leaked.
+func (cc *CorrelationCache) DeleteCache() {
+	close(cc.stopGc)
+}
+
+// Insert inserts a request into the correlation cache. After insertion, the
+// request can be correlated with responses via Correlate() or
+// CorrelateResponse()
+func (cc *CorrelationCache) Insert(req *RequestMessage) {
+	cc.mutex.Lock()
+	defer cc.mutex.Unlock()
+
+	correlationID := CorrelationID(req.GetCorrelationID())
+	if correlationID == 0 {
+		log.Warning("Correlation ID is zero")
+		return
+	}
+
+	if _, ok := cc.cache[correlationID]; ok {
+		log.Warning("Overwriting Kafka request message in correlation cache")
+	}
+
+	req.created = time.Now()
+	cc.cache[correlationID] = req
+}
+
+// Delete removes a request from the correlation cache. After deletion, the
+// request will no longer be correlated with responses. Typically this is done
+// when the request has been responded to.
+func (cc *CorrelationCache) Delete(req *RequestMessage) {
+	cc.mutex.Lock()
+	defer cc.mutex.Unlock()
+
+	correlationID := CorrelationID(req.GetCorrelationID())
+	if correlationID != 0 {
+		delete(cc.cache, correlationID)
+	}
+}
+
+// Correlate returns the request message with the matching correlation ID
+func (cc *CorrelationCache) Correlate(id CorrelationID) *RequestMessage {
+	cc.mutex.RLock()
+	defer cc.mutex.RUnlock()
+
+	req, _ := cc.cache[id]
+	return req
+}
+
+// CorrelateResponse extracts the correlation ID from the response message and
+// correlates the corresponding request
+func (cc *CorrelationCache) CorrelateResponse(res *ResponseMessage) *RequestMessage {
+	return cc.Correlate(CorrelationID(res.GetCorrelationID()))
+}
+
+func (cc *CorrelationCache) garbageCollector() {
+	for {
+		select {
+		case <-cc.stopGc:
+			return
+		}
+
+		expiryWatermark := time.Now().Add(requestLifetime)
+		log.WithField("expiryWatermark", expiryWatermark).
+			Debug("Running Kafka correlation cache garbage collector")
+
+		cc.mutex.Lock()
+		for correlationID, req := range cc.cache {
+			if req.created.After(expiryWatermark) {
```

Comment:
This will never be true. `req.created` is always <= `time.Now()`, so a fortiori it is always < `time.Now().Add(requestLifetime)`.
Shouldn't the logic be:
```
expiryWatermark := time.Now().Add(-requestLifetime)
...
if req.created.Before(expiryWatermark) {
```

Would it be possible to add a unit test for the GC logic? You might have to make the `requestLifetime` configurable.

Reply:
fixed